File: 0692-top-k-frequent-words/0692-top-k-frequent-words.cpp
class Solution {
public:
struct cmp{
    bool operator()(const pair<int,string>&a, const pair<int,string>&b){
        if(a.first!=b.first){
            return a.first>b.first;         // Min Heap of the First 
        }
        return b.second>a.second;           // Max heap on the Second one 
    }
};

    vector<string> topKFrequent(vector<string>& words, int k) {
        int n=words.size();
        
        // Priority Queue Declaration
        priority_queue<pair<int,string>,vector<pair<int,string>>,cmp> pq; 

        // Hashmap Declaration
        unordered_map<string,int> mp;

        // Loop for the Hashmap
        for(int i=0;i<n;i++){
            mp[words[i]]++;
        }  

        for(auto i:mp){
            string words=i.first;
            int freq=i.second;

            pair<int,string> curr={freq,words};

            if(pq.size()<k){
                pq.push(curr);
                continue;
            }

            if(curr.first<pq.top().first){
                continue;
            }

            if(curr.first==pq.top().first && curr.second>pq.top().second){
                continue;
            }
            pq.pop();
            pq.push(curr);
        }

        vector<string> res;

        while(!pq.empty()){
            res.push_back(pq.top().second);
            pq.pop();
        }

        reverse(res.begin(),res.end());

        return res; 
    }
};

File: 0692-top-k-frequent-words/0692-top-k-frequent-words.py
class Pair:
    def __init__(self,freq,word):
        self.first=freq
        self.second=word

    def __lt__(self,other):
        if(self.first!=other.first):
            return self.first<other.first
        return self.second>other.second


class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        n=len(words)

        # Priority Queue Declaration
        pq=[]

        # Hashmap Declaration
        mp={}

        # Loop for the Hashmap
        for i in range(n):
            mp[words[i]]=mp.get(words[i],0)+1

        for i in mp.items():
            words=i[0]
            freq=i[1]

            curr=Pair(freq,words)

            if(len(pq)<k):
                heapq.heappush(pq,curr)
                continue

            if(curr.first<pq[0].first):
                continue

            if(curr.first==pq[0].first and curr.second>pq[0].second):
                continue

            heapq.heappop(pq)
            heapq.heappush(pq,curr)

        res=[]

        while(pq):
            res.append(pq[0].second)
            heapq.heappop(pq)

        res.reverse()

        return res